pyccel/ast/bind_c.py

```diff
@@ -51,333 +59,445 @@ def __init__(self, *args, original_function, **kwargs):
-                        )
-
-#=======================================================================================
-def as_static_module(funcs, original_module):
-    """ Create the module contained in the bind_c_mod.f90 file
-    This is the interface between the c code and the fortran code thanks
-    to iso_c_bindings
+            return super().inout
+
+# =======================================================================================
+
+
+class BindCFunctionDefResult(FunctionDefResult):
+    """
+    Stores all the information necessary to expose a result to C code.
+
+    Results of a C-compatible function may need additional information
+    in order to fully construct the object. This class is mostly important
+    for array objects. These objects must describe not only the data, but also
+    meta-data. Namely the shape for the array in each dimension.
+    This information is stored in this class.
 
     Parameters
-    ==========
-    funcs : list of FunctionDef
-            All the functions which may be exposed to c
-    original_module : Module
-            The module being wrapped
+    ----------
+    var : Variable
+        The variable being returned (with a C-compatible type).
+
+    original_res_var : Variable
+        The variable which was returned by the function currently being wrapped
+        in a C-Fortran interface. This variable may have a type which is not
+        compatible with C.
+
+    scope : pyccel.parser.scope.Scope
+        The scope in which any arguments to the function should be declared.
+        This is used to create the shape and stride variables.
+
+    **kwargs : dict
+        See FunctionDefResult.
+
+    See Also
+    --------
+    pyccel.ast.core.FunctionDefResult
+        The class from which BindCFunctionDefResult inherits which
+        contains all details about the args and kwargs.
     """
-    funcs = [f for f in funcs if not f.is_private]
-    variables = [f for f in original_module.variables if not f.is_private]
-    imports = []
-    scope = Scope(used_symbols = original_module.scope.local_used_symbols.copy())
-    bind_c_funcs = [as_static_function_call(f, original_module, scope, imports = imports) for f in funcs]
-    bind_c_arrays = [wrap_module_array_var(v, scope, original_module) for v in variables if v.rank > 0]
-    if isinstance(original_module.name, AsName):
-        name = scope.get_new_name('bind_c_{}'.format(original_module.name.target))
-    else:
-        name = scope.get_new_name('bind_c_{}'.format(original_module.name))
-    return Module(name, (), bind_c_funcs+bind_c_arrays, imports = imports, scope=scope)
-
-#=======================================================================================
-def as_static_function_call(func, mod, mod_scope, name=None, imports = None):
+    __slots__ = ('_sizes', '_original_res_var')
+    _attribute_nodes = FunctionDefResult._attribute_nodes + \
+                        ('_var', '_sizes', '_original_res_var')
```

Comment:
`_var` is already contained in the tuple `FunctionDefResult._attribute_nodes`.

Reply:
Fixed acdc9945